source/common/upstream/outlier_detection_impl.cc

```diff
@@ -329,11 +329,15 @@ void DetectorImpl::checkHostForUneject(HostSharedPtr host, DetectorHostMonitorIm
     return;
   }
 
-  std::chrono::milliseconds base_eject_time =
-      std::chrono::milliseconds(runtime_.snapshot().getInteger(
-          "outlier_detection.base_ejection_time_ms", config_.baseEjectionTimeMs()));
-  ASSERT(monitor->numEjections() > 0);
-  if ((base_eject_time * monitor->numEjections()) <= (now - monitor->lastEjectionTime().value())) {
+  auto num_ejections = monitor->numEjections();
+  ASSERT_OR_LOG_AND(num_ejections > 0, return );
```

Comment:
drive by: all of these assertion changes feel unnecessary to me. If this can't happen why is a normal assert not ok? We don't do anything like this anywhere else.

Reply:
You are right this one could be.

But the specific problem we saw as that last_ejection_time was not populated, so it crashed. If we make that an ASSERT it will still crash. It would crash slightly more nicely if this were a RELEASE_ASSERT, but in this case I think it makes more sense to log the error and keep running.

Similar patterns are used a lot at Google -- this is visible in the Chromium source, using LOG(DFATAL), which would be used as:
```
  if (!last_ejection_time) {
   LOG(DFATAL) << "last_rejection_time is unset";
   return;
  }
```
That will assert-fail with a stack trace in debug builds, and log an error in production builds.

I was thinking it was less cumbersome to macro-ize the escape path too.

I'm also OK with the alternative of basically spelling this all out using Envoy macros:
```
#ifndef NDEBUG
  ASSERT(last_ejection_time);
#else 
  ENVOY_LOG_MISC(error, "last_ejection_time is unset");
  return;
#endif 
```

But I was thinking the pattern might be more generally useful so I proposed to macro-ize it. Happy to back that out if your skeptical of the usefulness :)

Reply:
Although I see where you are coming from, this change is IMO a large departure from our error handling philosophy specified in the style guide and we don't do anything like this anywhere else. If the path can't happen, we should be using a normal ASSERT. If it can happen, we should be adding a test for this path (which I think is impossible because it can't happen). So I would rather see the original bug fixed and we just keep using a normal assert and crash if there is a bug that needs to be fixed.

/wait-any

Reply:
Here's the stack-trace from where we saw this crash:
```
std::__u::__throw_bad_optional_access
std::__u::optional::value
Envoy::Upstream::Outlier::DetectorImpl::checkHostForUneject
Envoy::Upstream::Outlier::DetectorImpl::onIntervalTimer
Envoy::Upstream::Outlier::DetectorImpl::DetectorImpl::$_0::operator()
std::__u::__invoke
std::__u::__invoke_void_return_wrapper::__call
std::__u::__function::__default_alloc_func::operator()
std::__u::__function::__policy_invoker::__call_impl
std::__u::__function::__policy_func::operator()
std::__u::function::operator()
Envoy::Event::TimerImpl::TimerImpl::$_0::operator()
Envoy::Event::TimerImpl::TimerImpl::$_0::__invoke
event_process_active_single_queue
event_process_active
event_base_loop
Envoy::Server::InstanceImpl::run
Envoy::MainCommonBase::run
...
```
I don't know much about this code path but I was guessing (without data) this might have been during shutdown.

Reply:
OK. Can we fix the bug then and add a test for that?

Reply:
Definitely we want to fix the bug. The question is how to narrow it down given that we thought it couldn't happen.

RE not deviating from Envoy practices for error handling: SGTM.

But I claim there may be some value in having this being at least an ASSERT that's checked in prod. One thing is that my stack-trace might be misleading due to inlining, and so it'd be nice to get a file/line out of it rather than just a stack-trace.

So I'd propose that if nothing else we make it a RELEASE_ASSERT.

I also was thinking there may be value in letting the server continue to run, emitting error messages in production rather than crashing. That may tell us whether this is a one-time event or keeps happening on subsequent requests. If it just crashes (uncaught exception or assert) then we won't know that.

So I think these are the options
 1. convert to an ASSERT (but this won't help us diagnose issues in prod)
 2. convert to a RELEASE_ASSERT (better file/line diags in prod)
 3. let the process keep running after this occurs, whether using these macros or just
     sort of expanding them out for this one case.
 4. Just reason out how this can possibly occur.

Is option 2 a good compromise? 4 would be great too...I was just being lazy I guess and wanting to convince myself that this was the actual file/line.

Reply:
I would much rather just do (4) but I'm fine with (2) and a TODO/comment if that helps with debugging. I'm honestly very uncomfortable with (3) as we should just fix the bug and I agree that (1) won't help since this is a bug (I misunderstood initially).

Reply:
Cool...I committed a change to remove the new assert code and just use RELEASE_ASSERT.  I also looked a little deeper for option '4' and I need to better understand the threading model for this code.

Could eject() race with checkHostForUneject()? E.g. in thread 1:
```
eject calls host_.lock()->healthFlagSet(Host::HealthFlag::FAILED_OUTLIER_CHECK);
last_ejection_time is not set yet
```
then context-swtich to thread 2:
```
checkHostForUneject() is called
if (!host->healthFlagGet(Host::HealthFlag::FAILED_OUTLIER_CHECK)) { passes through
last_ejection_time is accessed
```
Is that plausible? If so we could mutex-protect a few member-vars...it's a little suspicious that there are some atomics here but no mutexes ensuring synchronization. It might be also possible to just re-order the statements in eject() to eliminate this, but it feels like then we should hold the ejection time in an atomic too.

....

Thinking about how this might be working a little more, this could also happen if there were two DetectorHostMonitorImpl which point to the same host, which is held in the structure as a `weak_ptr<Host>`. I think the use of `weak_ptr` here conveys a possibility that two different `DetectorHostMonitorImpl` instances point to the same Host, and thus the guard-statement that @junr03 referenced at the top of `checkHostForUneject()` may not be consistent with the DetectorHostMonitorImpl.

So...that seems to indicate there can only be one `DetectorHostMonitorImpl` per `Host`, but then I don't know why the weak_ptr.

And I'm still not sure whether in production two different threads may access `checkHostForUneject()` and `eject()`.

Reply:
@junr03 wrote this code originally and can likely take a look to see what the bug is. I can also take a look in the next couple of days if needed.

Reply:
I've been looking at this code a bit and I don't see how this can happen. Will keep looking a bit more. Do you have an actual core dump that you can look at?

All of the relevant operations happen on the main thread so there shouldn't be any multi-threaded eject/uneject logic. (In the consecutive failure case we post to the main thread to process the actual ejection.)

The use of weak_ptr is just to avoid a circular reference.

Reply:
I don't think we have a core file for this but I will check. 

In terms of reasoning about the root cause, if this class is entirely accessed from the main thread then my only other theory was if there were multiple DetectorImpl instances pointing to the same host, then you could get a status put on a host from one DetectorImpl instance, but another one doesn't have the ejection time populated. It doesn't sound like that's how the topology of these structures works.

Other than that I think it's random gamma rays and we could just drop this PR and see if it happens again, or merge it, in which case if it does happen again we might get slightly more info about it due to the filename & line number.

Reply:
I guess my preference would be to drop the PR until we see it again and hopefully get a core dump? (I'm not sure how much this PR would actually help debug this without a core dump anyway). WDYT?